src/NUnitFramework/framework/Internal/Execution/WorkItemBuilder.cs

```diff
@@ -56,42 +56,52 @@ static public WorkItem CreateWorkItem(ITest test, ITestFilter filter, bool recur
         /// <param name="filter">The filter to be used in selecting any child Tests.</param>
         /// <param name="debugger">An <see cref="IDebugger" /> instance.</param>
         /// <param name="recursive">True if child work items should be created and added.</param>
+        /// <param name="root">True iа work item needs to be created unconditionally, if false  null would be returned for tests that don't match the filter</param>
         /// <returns></returns>
-        internal static WorkItem CreateWorkItem(ITest test, ITestFilter filter, IDebugger debugger, bool recursive = false)
+        internal static WorkItem CreateWorkItem(ITest test, ITestFilter filter, IDebugger debugger, bool recursive = false, bool root = true)
```

Comment:
Similar to the `CountTestCases` this change makes it so we call `filter.Pass()` only on non-suite nodes. We use the presence of children as an indicator that the node needs to be created.

Reply:
Except for the case where this is the first invocation of the method and WI always needs to be created.